larreco/ClusterFinder/ClusterAna_module.cc

```diff
@@ -115,611 +138,471 @@ namespace cluster {
-      } // isCosmic
-      else {
-        float wght = 1.;
-        if (fTrackWeightOption == 1) wght = KE;
-        // accumulate statistics for muons and not-muons
-        if (PDG == 13) {
-          aveNuEP2mu += ep2 * wght;
-          numNuEP2mu += wght;
-        }
-        else {
-          aveNuEP2nm += ep2 * wght;
-          numNuEP2nm += wght;
-        }
-        if (PDG == 11) {
-          fNuKE_elec->Fill(KE, wght);
-          fNuE_elec->Fill(eff[ep2Plane], wght);
-          fNuP_elec->Fill(pur[ep2Plane], wght);
-          fNuEP2_elec->Fill(ep2, wght);
-          fNuEP2_KE_elec->Fill(KE, ep2, wght);
-        }
-        else if (PDG == 13) {
-          fNuKE_muon->Fill(KE, wght);
-          fNuE_muon->Fill(eff[ep2Plane], wght);
-          fNuP_muon->Fill(pur[ep2Plane], wght);
-          fNuEP2_muon->Fill(ep2, wght);
-          fNuEP2_KE_muon->Fill(KE, ep2, wght);
-        }
-        else if (PDG == 211) {
-          fNuKE_pion->Fill(KE, wght);
-          fNuE_pion->Fill(eff[ep2Plane], wght);
-          fNuP_pion->Fill(pur[ep2Plane], wght);
-          fNuEP2_pion->Fill(ep2, wght);
-          fNuEP2_KE_pion->Fill(KE, ep2, wght);
-        }
-        else if (PDG == 321) {
-          fNuKE_kaon->Fill(KE, wght);
-          fNuE_kaon->Fill(eff[ep2Plane], wght);
-          fNuP_kaon->Fill(pur[ep2Plane], wght);
-          fNuEP2_kaon->Fill(ep2, wght);
-          fNuEP2_KE_kaon->Fill(KE, ep2, wght);
+    } // fPrintLevel > 1
+
+    // Calculate Efficiency and Purity
+    ++fEventsProcessed;
+    for(auto& match : matches) {
+      if (match.mcpi == UINT_MAX) continue;
+      auto& mcp = (*mcps)[match.mcpi];
+      float TMeV = 1000 * (mcp.E() - mcp.Mass());
+      int indx = PDGCodeIndex(mcp.PdgCode());
+      if (indx < 0 || indx > 4) continue;
+      for(unsigned int plane = 0; plane < geom->Nplanes(); ++plane) {
+        if (match.mcpTruHitCount[plane] < 3) continue;
+        float eff = 0;
+        float pur = 0;
+        if (match.clsIndex[plane] != UINT_MAX) {
+          auto clsHits = fmch.at(match.clsIndex[plane]);
+          if (clsHits.size() > 0) {
+            eff = match.clsTruHitCount[plane] / match.mcpTruHitCount[plane];
+            pur = match.clsTruHitCount[plane] / (float)clsHits.size();
+          }
         }
-        else if (PDG == 2212) {
-          fNuKE_prot->Fill(KE, wght);
-          fNuE_prot->Fill(eff[ep2Plane], wght);
-          fNuP_prot->Fill(pur[ep2Plane], wght);
-          fNuEP2_prot->Fill(ep2, wght);
-          fNuEP2_KE_prot->Fill(KE, ep2, wght);
+        float effpur = eff * pur;
+        // accumulate
+        fEffSum[indx] += eff;
+        fPurSum[indx] += pur;
+        fEffPurSum[indx] += effpur;
+        ++fSum[indx];
+        // fill the histograms
+        if (indx == 0) {
+          fT_elec->Fill(TMeV);
+          fE_elec->Fill(eff);
+          fP_elec->Fill(pur);
+          fEP_elec->Fill(effpur);
+          fEP_T_elec->Fill(TMeV, effpur);
```

Comment:
If the histograms are included using the `ParticleInfo` structure suggested above, then this can become as simple as:
```.cpp
auto& info = info_for_particle[indx];
info.fT->Fill(TMeV);
info.fE->Fill(eff);
...
```